contracts/receivables/RealWorldReceivable.sol

```diff
@@ -0,0 +1,198 @@
+    mapping(uint256 => ReceivableInfo) public receivableInfoMapping;
+
+    // The status of the receivable.
+    enum Status {
+        Unpaid,
+        Paid,
+        PartiallyPaid
+    }
+
+    /**
+     * @dev Emitted when the owner of a receivable calls the declarePayment function
+     * @param from The address of the owner of the receivable
+     * @param to The address of the Pool that's paid by the receivable
+     * @param tokenId The ID of the receivable token
+     * @param amount The amount that was declared paid
+     */
+    event PaymentDeclared(
+        address indexed from,
+        address indexed to,
+        uint256 indexed tokenId,
+        uint256 amount
+    );
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @dev Initializer that sets the default admin and minter roles
+     */
+    function initialize() public initializer {
+        __ERC721_init("RealWorldReceivable", "RWR");
+        __ERC721Enumerable_init();
+        __ERC721URIStorage_init();
+        __ERC721Burnable_init();
+        __AccessControl_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MINTER_ROLE, msg.sender);
+    }
+
+    /**
+     * @dev Creates a new receivable token and assigns it to the recipient address
+     * @param recipient The address that will receive the new token
+     * @param paymentAddress The address that's expected to be paid for this receivable
+     * @param paymentToken The ERC20 token used to pay the receivable
+     * @param receivableAmount The total amount of the receivable
+     * @param maturityDate The date at which the receivable becomes due
+     * @param uri The URI of the metadata associated with the receivable
+     */
+    function safeMint(
+        address recipient,
+        address paymentAddress,
+        address paymentToken,
+        uint96 receivableAmount,
+        uint64 maturityDate,
+        string memory uri
+    ) public onlyRole(MINTER_ROLE) {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(recipient, tokenId);
+
+        ReceivableInfo memory receivableInfo = ReceivableInfo(
+            paymentAddress,
+            paymentToken,
+            receivableAmount,
+            0, // paidAmount
+            maturityDate
+        );
+        receivableInfoMapping[tokenId] = receivableInfo;
+
+        _setTokenURI(tokenId, uri);
+    }
+
+    /**
+     * @dev Declares payment for a receivable.
+     * Only the owner of the token can declare a payment.
+     * The payment method for the receivable must be Declarative.
+     * The receivable must not already be paid in full.
+     * Emits a `PaymentDeclared` event.
+     * @param tokenId The ID of the receivable token.
+     * @param paymentAmount The amount of payment being declared.
+     */
+    function declarePayment(uint256 tokenId, uint96 paymentAmount) public {
```

Comment:
I don't really like it. If a pool has multiple receivables active and do a partial payment on their loan, how would we decide which ones to pay off and which ones not to? It would be violating the receivables payment tracking and be a dangerous assumption for us to make. In my opinion we should enforce our partners to always use the receivables to make payments, especially once these receivables are tied closer to the pool.

Reply:
When I talked about "the loan", I was referring the amount our partners such as Arf borrowed from our pool, not the "loans" that are line entries for the receivable. The question is whether we should automatically mark all receivables paid off when the loan originated with the pool is fully paid off. If we do, to avoid changing the original contract, we will have to monitor the events and trigger changes here.

Reply:
We're talking about the same thing, that's exactly what I meant too. It seems like a dangerous assumption for us to write off receivables when we don't fully understand partners' operations. For example if a partner uses a different source of funds to pay off their loan, we'd erroneously mark receivables as paid when this isn't the case.

Reply:
On the flip side, the receivable's performance may look worse than it is if the partner misses uploading information after paying off the loan. You can counter argue that is an operational guideline that we should set with the partners - receivable payoff is independent from the pool loan payoff. Sounds good not to make change to this.